storage/mysql/log_storage.go

```diff
@@ -610,6 +610,17 @@ func (t *logTreeTX) GetSequencedLeafCount(ctx context.Context) (int64, error) {
 }
 
 func (t *logTreeTX) GetLeavesByIndex(ctx context.Context, leaves []int64) ([]*trillian.LogLeaf, error) {
+	if t.treeType == trillian.TreeType_LOG {
```

Comment:
Was copied from the checks in GetLeavesByRange. Is that out of date too?

Reply:
Dunno, I'm not up to speed on the differences.  @pavelkalinnikov  ?

Reply:
I think you're allowed to read past the published tree size for preordered logs but it seems to me that ByIndex and ByRange should be consistent.

Reply:
Correct, `GetLeavesByRange` can read past the tree size because LogSigner uses it to get new entries for building the tree. It does not use ByIndex though, but I think making it consistent with ByRange makes sense.

The `leaf < 0` check below is useful for both tree types though, maybe make it unconditional?

Reply:
Ah, I see.  Maybe we should have an `beyondAllowed bool` parameter to distinguish use of this storage function for an external gRPC request vs. internal use for the signer?

Reply:
I think it's a good idea.

Reply:
I think that API change should be done separately as this PR is trying to fix a client error mapping issue.

Reply:
Looks like the alternative idea is feasible. For instance, [GetLeavesByRange](https://github.com/google/trillian/blob/18b0fd3c43bfe044f161ad2abcd6e8b8acfcf88d/server/log_rpc_server.go#L516) could reduce `Count` passed in to storage, as it already has the tree size read from the same snapshot. Probably ByIndex could do the same. The advantage of such approach is that the storage interface stays the same.